packages/contracts/src/core/permission/PermissionManager.sol

```diff
@@ -454,6 +770,54 @@ abstract contract PermissionManager is Initializable {
         }
     }
 
+    /// @notice Function to check if the given address is ROOT.
+    /// @param _who The address to check for.
+    /// @return True if the given address is ROOT and otherwise false
+    function _isRoot(address _who) private returns (bool) {
+        return isGranted(address(this), _who, ROOT_PERMISSION_ID, msg.data);
+    }
+    
+    /// @notice Internal function to create a new permission.
+    /// @dev Requires the `ROOT_PERMISSION_ID` permission.
+    /// @param _where The address of the target contract for which `_who` receives permission.
+    /// @param _permissionIdOrSelector The permission hash or function selector used for this permission.
+    /// @param _owner The initial owner of this newly created permission.  
+    /// @param _whos The addresses of the target contracts for which `_who` receives permission.
+    function _createPermission(
+        address _where,
+        bytes32 _permissionIdOrSelector,
+        address _owner,
+        address[] calldata _whos
+    ) internal {
+        // let's say ROOT is the dao and only dao can call `createPermission`.
+        Permission storage permission = permissions[permissionHash(_where, _permissionIdOrSelector)];
+        if (permission.created) {
+            revert PermissionAlreadyCreated();
+        }
+
+        permission.created = true;
+        permission.owners[_owner] = uint8(15); // set flags to 00001111
+
+        if (_whos.length > 0) {
+            for (uint256 i = 0; i < _whos.length; i++) {
+                _grant(_where, _whos[i], _permissionIdOrSelector);
+            }
+        }
+
+        permission.grantCounter++;
+    }
+
+    /// @notice Internal function to check if this specific permission is frozen.
+    /// @param _permission Permission struct to check.
+    /// @return True if the permission is frozen and otherwise false
+    function _isPermissionFrozen(Permission storage _permission) private view returns (bool) {
+        return _permission.grantCounter == 0 && _permission.revokeCounter == 0;
```

Comment:
This is wrong as just because counters are 0, doesn't mean permission would be frozen. Owners can just remove themselves but that doesn't mean it's frozen.

You have to add extra check here - `permission.owners[address(1)].exists`.

Freezing basically means that one of the owner first had to call `addOwner(address(1))` and then, all owners must have removed themselves.

Reply:
Yeah, I have noted this down already.. this PR is draft :-)